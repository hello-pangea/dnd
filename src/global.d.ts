declare type AnimationFrameID = number;
declare type TimeoutID = number;
declare type IntervalID = number;

interface ProcessEnv {
  NODE_ENV: 'development' | 'production';
}

interface Process {
  env: ProcessEnv;
}

declare const process: Process;

declare interface Window {
  '__@hello-pangea/dnd-disable-dev-warnings'?: boolean;
}

// From the test folder we explicitly set typescript to load node.
// From the stories folder @storybook/react load NodeJS types.
// See: https://github.com/storybookjs/storybook/issues/12969#issue-733867804
// Loading node can causes issues similar to what is mention below
// See: https://github.com/Microsoft/TypeScript/issues/842
// In order to fix the problem we override some node types
declare namespace NodeJS {
  declare type Timeout = TimeoutID;
}
