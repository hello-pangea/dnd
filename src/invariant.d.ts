export declare class RbdInvariant {
  constructor(message: string);

  message: string;
}

export declare function invariant(condition: unknown, message?: string): void;
