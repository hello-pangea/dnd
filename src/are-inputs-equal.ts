/**
 * Original author: Alex Reardon
 * License: MIT
 * Repo: https://github.com/alexreardon/memoize-one
 */

function isEqual(first: unknown, second: unknown): boolean {
  if (first === second) {
    return true;
  }

  // Special case for NaN (NaN !== NaN)
  if (Number.isNaN(first) && Number.isNaN(second)) {
    return true;
  }

  return false;
}

export default function areInputsEqual(
  newInputs: readonly unknown[],
  lastInputs: readonly unknown[],
): boolean {
  // no checks needed if the inputs length has changed
  if (newInputs.length !== lastInputs.length) {
    return false;
  }
  // Using for loop for speed. It generally performs better than array.every
  // https://github.com/alexreardon/memoize-one/pull/59
  for (let i = 0; i < newInputs.length; i++) {
    if (!isEqual(newInputs[i], lastInputs[i])) {
      return false;
    }
  }
  return true;
}
