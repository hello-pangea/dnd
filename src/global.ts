interface ProcessEnv {
  NODE_ENV: 'development' | 'production';
}

interface Process {
  env: ProcessEnv;
}

declare global {
  const process: Process;

  interface Window {
    '__@react-forked/dnd-disable-dev-warnings'?: boolean;
  }
}

export {};
