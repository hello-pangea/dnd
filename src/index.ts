// Public types
export type {
  Announce,
  BeforeCapture,
  Direction,
  DraggableId,
  DraggableRubric,
  DragStart,
  DragUpdate,
  DraggableLocation,
  DroppableId,
  DropResult,
  Id,
  MovementMode,
  OnBeforeCaptureResponder,
  OnBeforeDragStartResponder,
  OnDragEndResponder,
  OnDragStartResponder,
  OnDragUpdateResponder,
  PreDragActions,
  ResponderProvided,
  SensorAPI,
  Sensor,
  SnapDragActions,
  TypeId,
  TryGetLock,
  TryGetLockOptions,
} from './types';

// DragDropContext
export { default as DragDropContext } from './view/drag-drop-context';
export type { DragDropContextProps } from './view/drag-drop-context';

// Draggable
export { default as Draggable } from './view/draggable';
export type {
  DraggableChildrenFn,
  DraggableProps,
  DraggableProvided,
  DraggableProvidedDraggableProps,
  DraggableProvidedDragHandleProps,
  DraggableStateSnapshot,
  DraggableStyle,
  DraggingStyle,
  DropAnimation,
  NotDraggingStyle,
} from './view/draggable/draggable-types';

// Droppable
export { default as Droppable } from './view/droppable';
export type {
  DroppableProps,
  DroppableProvided,
  DroppableProvidedProps,
  DroppableStateSnapshot,
} from './view/droppable/droppable-types';

// Sensors
export {
  useMouseSensor,
  useTouchSensor,
  useKeyboardSensor,
} from './view/use-sensor-marshal';

// Utils
export { resetServerContext } from './view/drag-drop-context';
