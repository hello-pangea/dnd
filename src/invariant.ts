/* eslint-disable no-restricted-syntax */
const isProduction: boolean = process.env.NODE_ENV === 'production';
const prefix = 'Invariant failed';

export class RbdInvariant extends Error {}

RbdInvariant.prototype.toString = function toString() {
  return this.message;
};

// A copy-paste of tiny-invariant but with a custom error type
// Throw an error if the condition fails
export function invariant(condition: false, message?: string): never;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function invariant(condition: any, message?: string): asserts condition;
export function invariant(
  condition: unknown,
  message?: string,
): asserts condition {
  if (condition) {
    return;
  }

  if (isProduction) {
    // In production we strip the message but still throw
    throw new RbdInvariant(prefix);
  } else {
    // When not in production we allow the message to pass through
    // *This block will be removed in production builds*
    throw new RbdInvariant(`${prefix}: ${message || ''}`);
  }
}
