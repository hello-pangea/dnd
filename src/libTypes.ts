export type RecursivePartial<T> = {
  [P in keyof T]?: RecursivePartial<T[P]>;
};
