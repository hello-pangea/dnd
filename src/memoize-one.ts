/**
 * Original author: Alex Reardon
 * License: MIT
 * Repo: https://github.com/alexreardon/memoize-one
 * Description: A memoization library which only remembers the latest invocation.
 */

import areInputsEqual from './are-inputs-equal';

export type EqualityFn<TFunc extends (...args: any[]) => any> = (
  newArgs: Parameters<TFunc>,
  lastArgs: Parameters<TFunc>,
) => boolean;

export interface MemoizedFn<TFunc extends (this: any, ...args: any[]) => any> {
  clear: () => void;
  (
    this: ThisParameterType<TFunc>,
    ...args: Parameters<TFunc>
  ): ReturnType<TFunc>;
}

// internal type
interface Cache<TFunc extends (this: any, ...args: any[]) => any> {
  lastThis: ThisParameterType<TFunc>;
  lastArgs: Parameters<TFunc>;
  lastResult: ReturnType<TFunc>;
}

export function memoizeOne<TFunc extends (this: any, ...newArgs: any[]) => any>(
  resultFn: TFunc,
  isEqual: EqualityFn<TFunc> = areInputsEqual,
): MemoizedFn<TFunc> {
  let cache: Cache<TFunc> | null = null;

  // breaking cache when context (this) or arguments change
  function memoized(
    this: ThisParameterType<TFunc>,
    ...newArgs: Parameters<TFunc>
  ): ReturnType<TFunc> {
    if (cache && cache.lastThis === this && isEqual(newArgs, cache.lastArgs)) {
      return cache.lastResult;
    }

    // Throwing during an assignment aborts the assignment: https://codepen.io/alexreardon/pen/RYKoaz
    // Doing the lastResult assignment first so that if it throws
    // the cache will not be overwritten
    const lastResult = resultFn.apply(this, newArgs);
    cache = {
      lastResult,
      lastArgs: newArgs,
      lastThis: this,
    };

    return lastResult;
  }

  // Adding the ability to clear the cache of a memoized function
  memoized.clear = function clear() {
    cache = null;
  };

  return memoized;
}
