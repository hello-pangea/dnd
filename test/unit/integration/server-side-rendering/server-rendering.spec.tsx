/**
 * @jest-environment node
 */
import React from 'react';
import { renderToString, renderToStaticMarkup } from 'react-dom/server';
import { invariant } from '../../../../src/invariant';
import { resetServerContext } from '../../../../src';
import App from '../util/app';
import invokeOnReactVersion from '../../../util/invoke-on-react-version';

let consoleWarnSpy: jest.SpiedFunction<typeof console.warn>;
let consoleErrorSpy: jest.SpiedFunction<typeof console.error>;
let consoleLogSpy: jest.SpiedFunction<typeof console.log>;

beforeEach(() => {
  // Reset server context between tests to prevent state being shared between them
  invokeOnReactVersion(['16', '17'], resetServerContext);

  consoleWarnSpy = jest.spyOn(console, 'warn');
  consoleErrorSpy = jest.spyOn(console, 'error');
  consoleLogSpy = jest.spyOn(console, 'log');
});

afterEach(() => {
  consoleWarnSpy.mockRestore();
  consoleErrorSpy.mockRestore();
  consoleLogSpy.mockRestore();
});

const expectConsoleNotCalled = () => {
  expect(consoleWarnSpy).not.toHaveBeenCalled();
  expect(consoleErrorSpy).not.toHaveBeenCalled();
  expect(consoleLogSpy).not.toHaveBeenCalled();
};

// Checking that the browser globals are not available in this test file
invariant(
  typeof window === 'undefined' && typeof document === 'undefined',
  'browser globals found in node test',
);

it('should support rendering to a string', () => {
  const result: string = renderToString(<App />);

  expect(result).toEqual(expect.any(String));
  expect(result).toMatchSnapshot();
  expectConsoleNotCalled();
});

it('should support rendering to static markup', () => {
  const result: string = renderToStaticMarkup(<App />);

  expect(result).toEqual(expect.any(String));
  expect(result).toMatchSnapshot();
  expectConsoleNotCalled();
});

// This test is unsuited for React 18+, which with useId can
// produce simultaneously unique and deterministic IDs
invokeOnReactVersion(['16', '17'], () => {
  it('should render identical content when resetting context between renders', () => {
    const firstRender = renderToString(<App />);
    const nextRenderBeforeReset = renderToString(<App />);
    expect(firstRender).not.toEqual(nextRenderBeforeReset);

    resetServerContext();
    const nextRenderAfterReset = renderToString(<App />);
    expect(firstRender).toEqual(nextRenderAfterReset);
    expectConsoleNotCalled();
  });
});
