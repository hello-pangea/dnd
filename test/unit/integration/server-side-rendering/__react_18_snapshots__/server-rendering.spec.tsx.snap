// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should support rendering to a string 1`] = `"<main><div data-rfd-droppable-id=\\"droppable\\" data-rfd-droppable-context-id=\\"0\\"><div data-rfd-draggable-context-id=\\"0\\" data-rfd-draggable-id=\\"0\\" tabindex=\\"0\\" role=\\"button\\" aria-describedby=\\"rfd-hidden-text-0-hidden-text-0\\" data-rfd-drag-handle-draggable-id=\\"0\\" data-rfd-drag-handle-context-id=\\"0\\" draggable=\\"false\\" data-is-dragging=\\"false\\" data-is-drop-animating=\\"false\\" data-is-combining=\\"false\\" data-is-combine-target=\\"false\\" data-is-clone=\\"false\\" data-testid=\\"0\\">item: <!-- -->0</div><div data-rfd-draggable-context-id=\\"0\\" data-rfd-draggable-id=\\"1\\" tabindex=\\"0\\" role=\\"button\\" aria-describedby=\\"rfd-hidden-text-0-hidden-text-0\\" data-rfd-drag-handle-draggable-id=\\"1\\" data-rfd-drag-handle-context-id=\\"0\\" draggable=\\"false\\" data-is-dragging=\\"false\\" data-is-drop-animating=\\"false\\" data-is-combining=\\"false\\" data-is-combine-target=\\"false\\" data-is-clone=\\"false\\" data-testid=\\"1\\">item: <!-- -->1</div><div data-rfd-draggable-context-id=\\"0\\" data-rfd-draggable-id=\\"2\\" tabindex=\\"0\\" role=\\"button\\" aria-describedby=\\"rfd-hidden-text-0-hidden-text-0\\" data-rfd-drag-handle-draggable-id=\\"2\\" data-rfd-drag-handle-context-id=\\"0\\" draggable=\\"false\\" data-is-dragging=\\"false\\" data-is-drop-animating=\\"false\\" data-is-combining=\\"false\\" data-is-combine-target=\\"false\\" data-is-clone=\\"false\\" data-testid=\\"2\\">item: <!-- -->2</div></div></main>"`;

exports[`should support rendering to static markup 1`] = `"<main><div data-rfd-droppable-id=\\"droppable\\" data-rfd-droppable-context-id=\\"0\\"><div data-rfd-draggable-context-id=\\"0\\" data-rfd-draggable-id=\\"0\\" tabindex=\\"0\\" role=\\"button\\" aria-describedby=\\"rfd-hidden-text-0-hidden-text-0\\" data-rfd-drag-handle-draggable-id=\\"0\\" data-rfd-drag-handle-context-id=\\"0\\" draggable=\\"false\\" data-is-dragging=\\"false\\" data-is-drop-animating=\\"false\\" data-is-combining=\\"false\\" data-is-combine-target=\\"false\\" data-is-clone=\\"false\\" data-testid=\\"0\\">item: 0</div><div data-rfd-draggable-context-id=\\"0\\" data-rfd-draggable-id=\\"1\\" tabindex=\\"0\\" role=\\"button\\" aria-describedby=\\"rfd-hidden-text-0-hidden-text-0\\" data-rfd-drag-handle-draggable-id=\\"1\\" data-rfd-drag-handle-context-id=\\"0\\" draggable=\\"false\\" data-is-dragging=\\"false\\" data-is-drop-animating=\\"false\\" data-is-combining=\\"false\\" data-is-combine-target=\\"false\\" data-is-clone=\\"false\\" data-testid=\\"1\\">item: 1</div><div data-rfd-draggable-context-id=\\"0\\" data-rfd-draggable-id=\\"2\\" tabindex=\\"0\\" role=\\"button\\" aria-describedby=\\"rfd-hidden-text-0-hidden-text-0\\" data-rfd-drag-handle-draggable-id=\\"2\\" data-rfd-drag-handle-context-id=\\"0\\" draggable=\\"false\\" data-is-dragging=\\"false\\" data-is-drop-animating=\\"false\\" data-is-combining=\\"false\\" data-is-combine-target=\\"false\\" data-is-clone=\\"false\\" data-testid=\\"2\\">item: 2</div></div></main>"`;
